internal/worker/removal/worker.go

```diff
@@ -62,8 +88,75 @@ func NewWorker(config Config) (worker.Worker, error) {
 }
 
 func (w *removalWorker) loop() (err error) {
-	<-w.catacomb.Dying()
-	return w.catacomb.Err()
+	watch, err := w.cfg.RemovalService.WatchRemovals()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	if err := w.catacomb.Add(watch); err != nil {
+		return errors.Capture(err)
+	}
+
+	ctx := w.catacomb.Context(context.Background())
+
+	timer := w.cfg.Clock.NewTimer(jobCheckMaxInterval)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case jobIDs := <-watch.Changes():
+			w.cfg.Logger.Infof(ctx, "got removal job changes: %v", jobIDs)
+			if err := w.processRemovalJobs(ctx); err != nil {
+				return errors.Capture(err)
+			}
+			timer.Reset(jobCheckMaxInterval)
+		case <-timer.Chan():
+			if err := w.processRemovalJobs(ctx); err != nil {
+				return errors.Capture(err)
+			}
+			timer.Reset(jobCheckMaxInterval)
+		}
+	}
+}
+
+// processRemovalJobs interrogates *all* current jobs scheduled for removal.
+// For each one whose scheduled start time has passed, we check to see if there
+// is an entry in our runner for it. If there is, it is already being processed
+// and we ignore it. Otherwise, it is commenced in a new runner.
+// This is safe due to the following conditions:
+// - This is the only method interrogating/adding workers to the runner.
+// - It is only invoked from cases in the main event loop, so is Goroutine safe.
+func (w *removalWorker) processRemovalJobs(ctx context.Context) error {
+	jobs, err := w.cfg.RemovalService.GetAllJobs(ctx)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	running := set.NewStrings(w.runner.WorkerNames()...)
+	log := w.cfg.Logger
+
+	for _, j := range jobs {
+		id := j.UUID.String()
+
+		if running.Contains(id) {
+			log.Debugf(ctx, "removal job %q already running", id)
```

Comment:
Is it? This will force workers that just died (since the call to `WorkerNames`) to wait until the next notification/timer fire.

`StartWorker` returns an _already exists_ error if there is a worker currently running with the same name.

If you are okay with that, then all I ask for is a comment.

Reply:
- If the worker died because it completed, it's fine. At this point we'll have deleted that job.
- If the worker died because it was incomplete, it's waiting for state to change and will try again in at most 30 seconds.
- If the worker died due to an error, we can afford to wait for up to 30 seconds for it to try again.

Reply:
They could be transitioning to removal simultaneously (different goroutines) but at the **time** of the acquisition of the lock to read the worker map they were running.

So, if you need to keep the worker running for some reason, this could be an issue.